src/stores/WidgetStore.ts

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2020 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Room } from "matrix-js-sdk/src/models/room";
+import { MatrixEvent } from "matrix-js-sdk/src/models/event";
+
+import { ActionPayload } from "../dispatcher/payloads";
+import { AsyncStoreWithClient } from "./AsyncStoreWithClient";
+import defaultDispatcher from "../dispatcher/dispatcher";
+import SettingsStore from "../settings/SettingsStore";
+import WidgetEchoStore from "../stores/WidgetEchoStore";
+import WidgetUtils from "../utils/WidgetUtils";
+import {SettingLevel} from "../settings/SettingLevel";
+import {WidgetType} from "../widgets/WidgetType";
+
+interface IState {}
+
+export interface IApp {
+    id: string;
+    type: string;
+    roomId: string;
+    eventId: string;
+    creatorUserId: string;
+    waitForIframeLoad?: boolean;
+    // eslint-disable-next-line camelcase
+    avatar_url: string; // MSC2765 https://github.com/matrix-org/matrix-doc/pull/2765
+}
+
+interface IRoomWidgets {
+    widgets: IApp[];
+    pinned: Record<string, boolean>;
+}
+
+// TODO consolidate WidgetEchoStore into this
+// TODO consolidate ActiveWidgetStore into this
+export default class WidgetStore extends AsyncStoreWithClient<IState> {
+    private static internalInstance = new WidgetStore();
+
+    private widgetMap = new Map<string, IApp>();
+    private roomMap = new Map<string, IRoomWidgets>();
+
+    private constructor() {
+        super(defaultDispatcher, {});
+
+        SettingsStore.watchSetting("Widgets.pinned", null, this.onPinnedWidgetsChange);
+        WidgetEchoStore.on("update", this.onWidgetEchoStoreUpdate);
+    }
+
+    public static get instance(): WidgetStore {
+        return WidgetStore.internalInstance;
+    }
+
+    private initRoom(roomId: string) {
+        if (!this.roomMap.has(roomId)) {
+            this.roomMap.set(roomId, {
+                pinned: {},
+                widgets: [],
+            });
+        }
+    }
+
+    protected async onReady(): Promise<any> {
+        this.matrixClient.on("RoomState.events", this.onRoomStateEvents);
```

Comment:
tbh this should probably be handled using `onAction` rather than off the client directly.

Reply:
Why? BreadcrumbsStore, OwnProfileStore listen directly like this

Reply:
ohgod.gif

tbh I think this is a bit of a bear trap for how stores work - let's leave it as written here for now and talk about it as a team eventually (ideally I'd like to kill off MatrixActions)